src/ChunkedDecoder.php

```diff
@@ -0,0 +1,197 @@
+<?php
+namespace React\Http;
+
+use Evenement\EventEmitter;
+use React\Stream\ReadableStreamInterface;
+use React\Stream\WritableStreamInterface;
+use React\Stream\Util;
+use Exception;
+
+/** @internal */
+class ChunkedDecoder extends EventEmitter implements ReadableStreamInterface
+{
+    const CRLF = "\r\n";
+
+    private $closed = false;
+    private $input;
+    private $buffer = '';
+    private $chunkSize = 0;
+    private $actualChunksize = 0;
+    private $chunkHeaderComplete = false;
+
+    public function __construct(ReadableStreamInterface $input)
+    {
+        $this->input = $input;
+
+        $this->input->on('data', array($this, 'handleData'));
+        $this->input->on('end', array($this, 'handleEnd'));
+        $this->input->on('error', array($this, 'handleError'));
+        $this->input->on('close', array($this, 'close'));
+    }
+
+
+    public function isReadable()
+    {
+        return ! $this->closed && $this->input->isReadable();
+    }
+
+    public function pause()
+    {
+        $this->input->pause();
+    }
+
+    public function resume()
+    {
+        $this->input->resume();
+    }
+
+    public function pipe(WritableStreamInterface $dest, array $options = array())
+    {
+        Util::pipe($this, $dest, $options);
+
+        return $dest;
+    }
+
+    public function close()
+    {
+        if ($this->closed) {
+            return;
+        }
+
+        $this->closed = true;
+
+        $this->input->close();
+
+        $this->emit('close');
+        $this->removeAllListeners();
+    }
+
+    /**
+     * Extracts the hexadecimal header and removes it from the given data string
+     *
+     * @param string $data - complete or incomplete chunked string
+     * @return string
+     */
+    private function handleChunkHeader($data)
+    {
+        $hexValue = strtok($this->buffer . $data, static::CRLF);
+        if ($this->isLineComplete($this->buffer . $data, $hexValue, strlen($hexValue))) {
+
+            if (dechex(hexdec($hexValue)) != $hexValue) {
+                $this->emit('error', array(new \Exception('Unable to identify ' . $hexValue . 'as hexadecimal number')));
+                $this->close();
+                return;
+            }
+
+            $this->chunkSize = hexdec($hexValue);
+            $this->chunkHeaderComplete = true;
+
+            $data = substr($this->buffer . $data, strlen($hexValue) + 2);
+            $this->buffer = '';
```

Comment:
This is *really* hard to keep track of. It looks like the buffer is modified in a couple of methods here? Can this buffering logic be simplified?

(see also above for `strpos()` etc.